Deserialize an entity from a protobuffer. (#6738)
googleapis/google-cloud-python: ndb/MIGRATION_NOTES.md

  is also a user-defined property named `key`. For an example, see the
  class docstring for `Model`.

## Bare Metal

One of the largest classes of differences comes from the use of the current
Datastore API, rather than the legacy App Engine Datastore. In general, for
users coding to the public interface, this won't be an issue, but users relying

> on pieces of the onstensibly private API that are exposed to the bare metal of

Reviewer: ```suggestion
on pieces of the ostensibly private API that are exposed to the bare metal of
```

the original datastore implementation will have to rewrite those pieces.

> Specifically, any function or method that dealt directly with protobuffers will

Reviewer: ```suggestion
Specifically, any function or method that dealt directly with protocol buffers will
```

I don't think "protobuffers" is a thing?

no longer work. Datastore, itself, has moved to gRPC and protobuffers

> themselves are significantly different from legacy Datastore.  Additionally,

Reviewer: I would rewrite this to just saw the `.protobuf` definitions have changed quite a bit from the barely-public API used by App Engine to the [public one][1] (also drop all usage of "protobuffers").

[1]: https://github.com/googleapis/googleapis/tree/master/google/datastore/v1

Author: ab2f965

this version of NDB makes less direct use of protobuffers and relies more

> heavily on Datastore for parsing and generating protobuffers, which is a

Reviewer: ISTM you mean `google-cloud-datastore` rather than the actual backend service "Datastore" here.

Author: ab2f965

significant internal refactoring.

- `ModelAdapter` is no longer used. In legacy NDB, this was passed to Datastore
  so that calls to Datastore RPCs could yield NDB entities directly from
  Datastore RPC calls. AFAIK, Datastore no longer accepts an adapter for
  adpating entities. At any rate, we no longer do it that way.
- `Property._db_get_value` is no longer used. It worked directly with Datastore

>   protobuffers, work which is now delegated to Datastore.

Author: @dhermes You mention above dropping use of "protobuffers"--how should I refer to the `p` argument of the new deprecated `Property._db_get_value`?

Reviewer: If it's deprecated, do we need to document it?

Reviewer: But also, you can just say "protocol buffer" every time you were using "protobuffer".

Author: I think it's similar status to a lot of stuff in this file: probably not an issue but worth mentioning in case someone was overriding `Property._db_get_value` for some ill-advised reason. I get the impression that best practices for NDB evolved over time. There's an element of archaeology to this work.

Reviewer: :+1:

Author: ab2f965

- `Model._deserialize` is no longer used. It worked directly with protobuffers,
  so wasn't really salvageable. Unfortunately, there were comments indicating

>   was overridden by subclasses. Hopefully this isn't broadly the case.

Reviewer: ```suggestion
  it was overridden by subclasses. Hopefully this isn't broadly the case.
```

  is also a user-defined property named `key`. For an example, see the
  class docstring for `Model`.

## Bare Metal

One of the largest classes of differences comes from the use of the current
Datastore API, rather than the legacy App Engine Datastore. In general, for
users coding to the public interface, this won't be an issue, but users relying
on pieces of the ostensibly private API that are exposed to the bare metal of
the original datastore implementation will have to rewrite those pieces.
Specifically, any function or method that dealt directly with protocol buffers
will no longer work. The Datastore `.protobuf` definitions have changed
significantly from the barely public API used by App Engine to the current
published API. Additionally, this version of NDB mostly delegates to
`google.cloud.datastore` for parsing data returned by RPCs, which is a
significant internal refactoring.

- `ModelAdapter` is no longer used. In legacy NDB, this was passed to the 
  Datastore RPC client so that calls to Datastore RPCs could yield NDB entities
  directly from Datastore RPC calls. AFAIK, Datastore no longer accepts an

>   adapter for adpating entities. At any rate, we no longer do it that way.

Reviewer: ```suggestion
  adapter for adapting entities. At any rate, we no longer do it that way.
```
